### PR TITLE
Verify VM host resources are reclaimed correctly

### PR DESCRIPTION
**Pull request by enescakir**
- **Fix used_hugepages_1g calculation after VM host reboot**
  We verify and recalculate hugepages after rebooting the VM host by
  summing SPDK hugepages and VM hugepages.

  When a burstable VM is allocated on the host,
  `vm_host.used_hugepages_1g` is incremented by the hugepage usage of the
  slice (8g), not the burstable VM itself (4g). Therefore, if slicing is
  enabled on the host, we should sum `vm_host_slice.total_memory_gib`
  instead of VM memory.

- **Split vm host wait and spec installation into two separate labels**
  They serve very different purposes, so separating them will improve
  clarity

- **Wait for postgres resources to be fully destroyed in E2E tests**

- **Verify VM host resources are reclaimed correctly**
  Recently, we had a production bug where the destroy label for a VM ran
  twice, causing the VM host resources to be reclaimed twice. This
  resulted in negative resource values and failed constraints. Adding a
  check in the E2E tests will help verify this case. While it may not
  catch every edge case, it will at least cover the basic scenario.
  

----

> [!IMPORTANT]
> Fix hugepages calculation and improve VM host resource reclamation and verification logic, with updated tests for resource cleanup.
> 
>   - **Behavior**:
>     - Fix `used_hugepages_1g` calculation in `host_nexus.rb` to sum `vm_host_slice.total_memory_gib` if slicing is enabled.
>     - Add checks in `verify_resources_reclaimed` in `hetzner_server.rb` to ensure `used_cores`, `used_hugepages_1g`, and `available_storage_gib` are zero after resource reclamation.
>     - Split `wait_setup_host` into `wait_setup_host` and `install_integration_specs` in `hetzner_server.rb` for clarity.
>   - **Tests**:
>     - Add `wait_resources_destroyed` in `ha_postgres_resource.rb` and `postgres_resource.rb` to ensure resources are fully destroyed before proceeding.
>     - Add tests for `verify_resources_reclaimed` in `hetzner_server_spec.rb` to check for proper resource cleanup.
>     - Update tests in `ha_postgres_resource_spec.rb` and `postgres_resource_spec.rb` to reflect new resource destruction logic.
>   - **Misc**:
>     - Minor refactoring in `hetzner_server.rb` to improve code clarity and separation of concerns.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ubicloud%2Fubicloud&utm_source=github&utm_medium=referral)<sup> for 4c1987d20bca1bf9fd27d930a7874e4c03473f62. You can [customize](https://app.ellipsis.dev/ubicloud/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by enescakir** [comment]
I ran E2E, and `vm_host.used_hugepages_1g` was `-2`, so it failed. @pykello, do you have any idea how this could have happened?

https://github.com/ubicloud/ubicloud/actions/runs/18201965169/job/51822879309

**Review by jeremyevans — COMMENTED** [review]
Ruby-wise, the code reads fine, but since the E2E tests don't pass, something isn't right. To debug the issue, it might be worth recording the VmHost values in multiple labels in the test prog. For example, might be worth testing the values decrease as expected after the VM is added, and remain there until right before the VM is destroyed.  This is assuming @pykello doesn't already know the cause.

**Comment by enescakir** [comment]
I printed all DB queries with `ENV["DATABASE_LOGGER_LEVEL"] ||= "debug"`

https://github.com/ubicloud/ubicloud/actions/runs/18225295943/job/51894841338

**Comment by enescakir** [comment]
Then I grep used_hugepages_1g` updates with `grep -E "UPDATE \"vm_host\".*used_hugepages_1g" 0_run-ci.txt`

**Comment by enescakir** [comment]
There are 42 queries.

20 of them increase the value by 8 when we allocate a VM on the host
```
2025-10-03T15:24:18.9879263Z 15:00:26 respirate.2 | I, [2025-10-03T15:00:26.043211 #5193]  INFO -- : (0.000449s) UPDATE "vm_host" SET "used_cores" = ("used_cores" + $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" + $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:19.0050004Z 15:00:26 respirate.1 | I, [2025-10-03T15:00:26.976300 #5195]  INFO -- : (0.000345s) UPDATE "vm_host" SET "used_cores" = ("used_cores" + $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" + $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:19.0189178Z 15:00:28 respirate.1 | I, [2025-10-03T15:00:28.839703 #5195]  INFO -- : (0.000409s) UPDATE "vm_host" SET "used_cores" = ("used_cores" + $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" + $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:20.6748317Z 15:13:56 respirate.2 | I, [2025-10-03T15:13:56.311867 #5193]  INFO -- : (0.000638s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:20.6769485Z 15:13:56 respirate.2 | I, [2025-10-03T15:13:56.342547 #5193]  INFO -- : (0.000468s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:20.6855434Z 15:13:56 respirate.1 | I, [2025-10-03T15:13:56.497586 #5195]  INFO -- : (0.001313s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:20.6968169Z 15:13:56 respirate.1 | I, [2025-10-03T15:13:56.594350 #5195]  INFO -- : (0.000358s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:20.7064038Z 15:13:57 respirate.2 | I, [2025-10-03T15:13:57.712527 #5193]  INFO -- : (0.000374s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:20.9632058Z 15:13:59 respirate.2 | I, [2025-10-03T15:13:59.234362 #5193]  INFO -- : (0.000393s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:21.0440464Z 15:13:59 respirate.1 | I, [2025-10-03T15:13:59.581035 #5195]  INFO -- : (0.000382s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:21.0798723Z 15:13:59 respirate.2 | I, [2025-10-03T15:13:59.745514 #5193]  INFO -- : (0.000166s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:25.3659614Z 15:14:25 respirate.2 | I, [2025-10-03T15:14:25.522026 #5193]  INFO -- : (0.000485s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:25.9411869Z 15:15:34 respirate.1 | I, [2025-10-03T15:15:34.829411 #5195]  INFO -- : (0.000402s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:25.9512862Z 15:15:35 respirate.1 | I, [2025-10-03T15:15:35.844985 #5195]  INFO -- : (0.000287s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:26.1718167Z 15:16:14 respirate.2 | I, [2025-10-03T15:16:14.065633 #5193]  INFO -- : (0.000356s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:26.1890218Z 15:16:14 respirate.2 | I, [2025-10-03T15:16:14.923346 #5193]  INFO -- : (0.000381s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:26.3678049Z 15:16:24 respirate.1 | I, [2025-10-03T15:16:24.881520 #5195]  INFO -- : (0.000415s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:30.5016289Z 15:16:58 respirate.1 | I, [2025-10-03T15:16:58.244915 #5195]  INFO -- : (0.000156s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:30.5314552Z 15:17:01 respirate.2 | I, [2025-10-03T15:17:01.278393 #5193]  INFO -- : (0.000220s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:34.3792461Z 15:18:30 respirate.2 | I, [2025-10-03T15:18:30.879213 #5193]  INFO -- : (0.000441s) UPDATE "vm_host" SET "used_hugepages_1g" = ("used_hugepages_1g" + $1::int4) WHERE ("id" = $2); [8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
```

20 decrease it by 8 when we deallocate a VM
```
2025-10-03T15:24:20.5994710Z 15:13:40 respirate.2 | I, [2025-10-03T15:13:40.983656 #5193]  INFO -- : (0.000583s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:20.6121961Z 15:13:41 respirate.2 | I, [2025-10-03T15:13:41.279034 #5193]  INFO -- : (0.000750s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:20.6197354Z 15:13:42 respirate.2 | I, [2025-10-03T15:13:42.085284 #5193]  INFO -- : (0.000579s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:26.1624561Z 15:16:13 respirate.2 | I, [2025-10-03T15:16:13.982984 #5193]  INFO -- : (0.000514s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:30.5601087Z 15:17:04 respirate.2 | I, [2025-10-03T15:17:04.177010 #5193]  INFO -- : (0.000196s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:31.1780500Z 15:17:08 respirate.1 | I, [2025-10-03T15:17:08.863126 #5195]  INFO -- : (0.000437s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:32.2283612Z 15:17:19 respirate.1 | I, [2025-10-03T15:17:19.937996 #5195]  INFO -- : (0.000477s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:34.0914772Z 15:17:36 respirate.2 | I, [2025-10-03T15:17:36.998863 #5193]  INFO -- : (0.000373s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:34.1987834Z 15:17:47 respirate.1 | I, [2025-10-03T15:17:47.950112 #5195]  INFO -- : (0.000286s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:34.3187298Z 15:18:10 respirate.1 | I, [2025-10-03T15:18:10.791840 #5195]  INFO -- : (0.000458s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:34.8677237Z 15:20:52 respirate.2 | I, [2025-10-03T15:20:52.531266 #5193]  INFO -- : (0.000302s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:34.8730642Z 15:20:52 respirate.1 | I, [2025-10-03T15:20:52.747809 #5195]  INFO -- : (0.000204s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:34.8823257Z 15:20:54 respirate.2 | I, [2025-10-03T15:20:54.219123 #5193]  INFO -- : (0.000137s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:34.8920076Z 15:20:55 respirate.2 | I, [2025-10-03T15:20:55.678098 #5193]  INFO -- : (0.000309s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:35.0623539Z 15:23:16 respirate.2 | I, [2025-10-03T15:23:16.898885 #5193]  INFO -- : (0.000348s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:35.0709580Z 15:23:17 respirate.1 | I, [2025-10-03T15:23:17.714635 #5195]  INFO -- : (0.000410s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:35.0769770Z 15:23:17 respirate.2 | I, [2025-10-03T15:23:17.927450 #5193]  INFO -- : (0.000382s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:35.1481915Z 15:23:48 respirate.1 | I, [2025-10-03T15:23:48.187508 #5195]  INFO -- : (0.000312s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:35.1553009Z 15:23:49 respirate.2 | I, [2025-10-03T15:23:49.645771 #5193]  INFO -- : (0.000455s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:35.1647758Z 15:23:49 respirate.1 | I, [2025-10-03T15:23:49.836954 #5195]  INFO -- : (0.000414s) UPDATE "vm_host" SET "used_cores" = ("used_cores" - $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" - $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
```

One is the initial query that sets the SPDK hugepages. 
```
2025-10-03T15:24:18.9335073Z 14:59:21 respirate.1 | I, [2025-10-03T14:59:21.788257 #5195]  INFO -- : (0.000339s) UPDATE "vm_host" SET "total_hugepages_1g" = $1::int4, "used_hugepages_1g" = $2::int4 WHERE ("id" = $3); [249, 2, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
```

This one looks suspicious.
```
2025-10-03T15:24:19.9648310Z 15:08:40 respirate.1 | I, [2025-10-03T15:08:40.388739 #5195]  INFO -- : (0.000241s) UPDATE "vm_host" SET "used_hugepages_1g" = $1::int4 WHERE ("id" = $2); [22, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]

```

**Why we set it to 22?**

**Comment by enescakir** [comment]
I looked at the lines around it.

```
2025-10-03T15:24:19.9646495Z 15:08:40 respirate.1 | I, [2025-10-03T15:08:40.387231 #5195]  INFO -- : (0.000242s) SELECT * FROM "vm_host" WHERE "id" = '08d49353-cd3c-8f71-ade3-5c9b674948e6'                                                                                                                                              2025-10-03T15:24:19.9647135Z 15:08:40 respirate.1 | I, [2025-10-03T15:08:40.387660 #5195]  INFO -- : (0.000195s) SELECT * FROM "spdk_installation" WHERE ("spdk_installation"."vm_host_id" = $1); ["08d49353-cd3c-8f71-ade3-5c9b674948e6"]                                                                                                2025-10-03T15:24:19.9647663Z 15:08:40 respirate.1 | I, [2025-10-03T15:08:40.387970 #5195]  INFO -- : (0.000191s) SELECT * FROM "vm" WHERE ("vm"."vm_host_id" = $1); ["08d49353-cd3c-8f71-ade3-5c9b674948e6"]                                                                                                                              2025-10-03T15:24:19.9648310Z 15:08:40 respirate.1 | I, [2025-10-03T15:08:40.388739 #5195]  INFO -- : (0.000241s) UPDATE "vm_host" SET "used_hugepages_1g" = $1::int4 WHERE ("id" = $2); [22, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]                                                                                                      2025-10-03T15:24:19.9649054Z 15:08:40 respirate.1 | {"strand_hopped":{"duration":0.388854552,"from":"Vm::HostNexus.verify_hugepages","to":"Vm::HostNexus.start_slices"},"message":"hopped","time":"2025-10-03 15:08:40 +0000","thread":"vh13a96myd7j7nqhns6v7954ec"}
```

After the initial setup, we verify the hugepages and update their value again. The `total_vm_mem_gib` is 20, and `spdk_hugepages` is 2, so we set it to 22.

https://github.com/ubicloud/ubicloud/blob/67261e7cbf0695933a9da93beee588a606c8c6fd/prog/vm/host_nexus.rb#L256

**Comment by enescakir** [comment]
We allocate 3 vms on the host, so it should be `3 * 8 + 2 = 26` not 22 @pykello 

```
2025-10-03T15:24:18.9335073Z 14:59:21 respirate.1 | I, [2025-10-03T14:59:21.788257 #5195]  INFO -- : (0.000339s) UPDATE "vm_host" SET "total_hugepages_1g" = $1::int4, "used_hugepages_1g" = $2::int4 WHERE ("id" = $3); [249, 2, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]

2025-10-03T15:24:18.9879263Z 15:00:26 respirate.2 | I, [2025-10-03T15:00:26.043211 #5193]  INFO -- : (0.000449s) UPDATE "vm_host" SET "used_cores" = ("used_cores" + $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" + $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:19.0050004Z 15:00:26 respirate.1 | I, [2025-10-03T15:00:26.976300 #5195]  INFO -- : (0.000345s) UPDATE "vm_host" SET "used_cores" = ("used_cores" + $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" + $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
2025-10-03T15:24:19.0189178Z 15:00:28 respirate.1 | I, [2025-10-03T15:00:28.839703 #5195]  INFO -- : (0.000409s) UPDATE "vm_host" SET "used_cores" = ("used_cores" + $1::int4), "used_hugepages_1g" = ("used_hugepages_1g" + $2::int4) WHERE ("id" = $3); [1, 8, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]

2025-10-03T15:24:19.9648310Z 15:08:40 respirate.1 | I, [2025-10-03T15:08:40.388739 #5195]  INFO -- : (0.000241s) UPDATE "vm_host" SET "used_hugepages_1g" = $1::int4 WHERE ("id" = $2); [22, "08d49353-cd3c-8f71-ade3-5c9b674948e6"]
```

File: prog/vm/host_nexus.rb [hunk 1]
```diff
@@ -245,10 +245,14 @@ def os_supports_slices?(os_version)
     total_hugepages = Integer(total_hugepages_match.captures.first)
     free_hugepages = Integer(free_hugepages_match.captures.first)
 
-    spdk_hugepages = vm_host.spdk_installations.sum { |i| i.hugepages }
+    spdk_hugepages = vm_host.spdk_installations.sum(&:hugepages)
     fail "Used hugepages exceed SPDK hugepages" unless total_hugepages - free_hugepages <= spdk_hugepages
 
-    total_vm_mem_gib = vm_host.vms.sum { |vm| vm.memory_gib }
+    total_vm_mem_gib = if vm_host.accepts_slices
```
Comment (jeremyevans) [inline prog/vm/host_nexus.rb:266]: Do we ever have a host that has some VMs in slices and some not in slices? If so, does this code need adjustment to handle that case? If not, this seems fine.

**Review by jeremyevans — APPROVED** [review]

File: prog/vm/host_nexus.rb (hunk 1 above)
Comment (enescakir) [reply prog/vm/host_nexus.rb:266]: As I remember, we decided not to mix slice and non-slice VMs on the same host. Is that right, @pykello?

Comment (pykello) [reply prog/vm/host_nexus.rb:266]: Yes, we're not mixing slice & non-slice on the same host.
